src/neural/network_mux.cc

```diff
@@ -62,10 +62,8 @@ class MuxingComputation : public NetworkComputation {
   }
 
   void NotifyReady() {
-    {
-      std::unique_lock<std::mutex> lock(mutex_);
-      dataready_ = true;
-    }
+    std::unique_lock<std::mutex> lock(mutex_);
+    dataready_ = true;
```

Comment:
why not make this atomic?

Reply:
dataready_cv_.notify_one() below notifies the waiter to continue and the problem would be back, this thread can be suspended in notify_one and the waiter thread given priority to continue destroying the object before notify_one finishes in the other thread.